app/routes/__init__.py

```diff
@@ -137,6 +146,45 @@ async def raster_tile_cache_version_dependency(
     return dataset, version
 
 
+async def cog_asset_dependency(
+    request: Request,
+    dataset: Optional[COGDatasets] = Query(None, description=COGDatasets.__doc__),  # type: ignore
+    version: Optional[str] = Query(
+        None, description="Data API dataset version.", regex=VERSION_REGEX_NO_LATEST
+    ),
+    url: Optional[str] = Query(
+        None,
+        description="Dataset path. This needs to be set if `dataset` and `version` query parameters for a Data API dataset are not set.",
+    ),
+) -> Optional[str]:
+
+    if dataset is None and version is None and url is None:
+        raise HTTPException(
+            status.HTTP_400_BAD_REQUEST,
+            "Need to pass either `url` or `dataset` and `version` pair for Data API dataset.",
+        )
+
+    if dataset and version and url:
```

Comment:
Very minor quibble - you are not giving and error in the case where they pass dataset and URL only, or version and URL only.  In that case, you will just use the URL, which is probably ok.